### PR TITLE
OP-1184 | Add new inventory for all products or for specific product

### PR DESCRIPTION
**Pull request by ArnaudFonzam**
See OP-1184.

### CODE REVIEW

File: bundle/language_en.properties [hunk 1]
```diff
@@ -663,19 +663,39 @@ angal.hospital.thevisitinghourvaluesmustbeintherange0to24.msg
 angal.hospital.visitduration.txt                                                                       = Visit duration
 angal.hospital.visitendhour.txt                                                                        = Visit end hour
 angal.hospital.visitstarthour.txt                                                                      = Visit start hour
+angal.inventory.addlotcodebeforeenterthecost                                                           = Add lot code before enter the cost.
```
Comment (dbmalkovsky) [inline bundle/language_en.properties:666]:
```suggestion
angal.inventory.addlotcodebeforeenteringthecost.msg                                                       = Add lot code before entering the cost.
```

File: bundle/language_en.properties [hunk 2]
```diff
@@ -663,19 +663,39 @@ angal.hospital.thevisitinghourvaluesmustbeintherange0to24.msg
...
 angal.inventory.allproduct.txt                                                                         = All products
 angal.inventory.canceled                                                                               = Canceled
 angal.inventory.cannotcreateanotherinventorywithstatusdraft.msg                                        = Can not create another inventory with status draft.
+angal.inventory.cannotsavelot                                                                          = Can't save lot. Please verify lot informations and retry.
```
Comment (dbmalkovsky) [inline bundle/language_en.properties:670]:
```suggestion
angal.inventory.cannotsavelot.msg                                                                     = Cannot save the lot. Please verify lot information and retry.
```

File: bundle/language_en.properties [hunk 3]
```diff
@@ -663,19 +663,39 @@ angal.hospital.thevisitinghourvaluesmustbeintherange0to24.msg
...
 angal.inventory.draft                                                                                  = Draft
 angal.inventory.inventorybrowser.title                                                                 = Inventory Browser
+angal.inventory.mustenterareference.msg                                                                = You must enter a reference.
 angal.inventory.neweditinventory.title                                                                 = Edit an inventory
+angal.inventory.notdateinfuture.msg                                                                    = Date in future is not allowed !
```
Comment (dbmalkovsky) [inline bundle/language_en.properties:675]:
```suggestion
angal.inventory.notdateinfuture.msg                                                                    = A date in the future is not allowed
```

File: bundle/language_en.properties [hunk 4]
```diff
@@ -663,19 +663,39 @@ angal.hospital.thevisitinghourvaluesmustbeintherange0to24.msg
...
+angal.inventory.noproduct.msg                                                                          = Product not found!
```
Comment (dbmalkovsky) [inline bundle/language_en.properties:676]:
```suggestion
angal.inventory.noproduct.msg                                                                          = Product is not found/
```

File: bundle/language_en.properties [hunk 5]
```diff
@@ -663,19 +663,39 @@ angal.hospital.thevisitinghourvaluesmustbeintherange0to24.msg
...
+angal.inventory.referencealreadyused.msg                                                               = Reference is already used.
+angal.inventory.savesucces.msg                                                                         = Inventory saved. 
+angal.inventory.error.msg                                                                              = Error during  some products's recording
+angal.inventoryrow.cannotchangethepriceofproductwithoutlot.msg                                            = You can not update the unit price of product without lot.
```
Comment (dbmalkovsky) [inline bundle/language_en.properties:680]:
```suggestion
angal.inventoryrow.cannotchangethepriceofproductwithoutlot.msg                                         = You cannot update the unit price of a product without a lot.
```

File: bundle/language_en.properties [hunk 6]
```diff
@@ -663,19 +663,39 @@ angal.hospital.thevisitinghourvaluesmustbeintherange0to24.msg
...
+angal.inventoryrow.doyoureallywanttodeletethisinventoryrow.msg                                         = Do you really want to delete this inventory row?
+angal.inventoryrow.pleaseselectonlyoneinventoryrow.msg                                                 = Please select only one inventory row.
+angal.inventoryoperation.save.error.msg                                                                = Inventory not saved
```
Comment (dbmalkovsky) [inline bundle/language_en.properties:683]:
```suggestion
angal.inventoryoperation.save.error.msg                                                                = Inventory not saved.
```

File: bundle/language_en.properties [hunk 7]
```diff
@@ -663,19 +663,39 @@ angal.hospital.thevisitinghourvaluesmustbeintherange0to24.msg
...
 angal.inventory.nextarrow.btn                                                                          = >
 angal.inventory.arrowprevious.btn                                                                      = <
 angal.inventory.product.col                                                                            = Products
+angal.inventory.savelotwithnullprice                                                                   = Save lot with null price ?
```
Comment (dbmalkovsky) [inline bundle/language_en.properties:687]:
```suggestion
angal.inventory.savelotwithnullprice.msg                                                               = Save the lot without a price?
```

File: bundle/language_en.properties [hunk 8]
```diff
@@ -663,19 +663,39 @@ angal.hospital.thevisitinghourvaluesmustbeintherange0to24.msg
...
+angal.inventory.selectmedical                                                                          = Select Medical
```
Comment (dbmalkovsky) [inline bundle/language_en.properties:688]:
```suggestion
angal.inventory.selectmedical.msg                                                                      = Select a medical.
```

File: bundle/language_en.properties [hunk 9]
```diff
@@ -854,6 +874,7 @@ angal.medicalstock.multiplecharging.preparationdate
 angal.medicalstock.multiplecharging.qtypacket                                                          = Qty/Packet
 angal.medicalstock.multiplecharging.referencenumberabb                                                 = Reference No.
 angal.medicalstock.multiplecharging.selectedlot                                                        = Selected Lot
+angal.medicalstock.multiplecharging.selectmedical.title                                                = Choose a medical
```
Comment (dbmalkovsky) [inline bundle/language_en.properties:877]:
```suggestion
angal.medicalstock.multiplecharging.selectmedical.title                                                = Choose a Medical
```
titles use caps

File: src/main/java/org/isf/medicalinventory/gui/InventoryBrowser.java [hunk 10]
```diff
@@ -274,10 +273,21 @@ private GoodDateChooser getJCalendarTo() {
...
-					dateTo = newDate.atTime(LocalTime.MAX);
+				dateTo = jCalendarTo.getDateStartOfDay();
+				if (inventoryList != null) {
+					inventoryList.clear();
 				}
+				InventoryBrowsingModel inventoryModel = new InventoryBrowsingModel();
+				TOTAL_ROWS = inventoryModel.getRowCount();
+				START_INDEX = 0;
+				previous.setEnabled(false);
+				if (TOTAL_ROWS <= PAGE_SIZE){
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/medicalinventory/gui/InventoryBrowser.java:284]:
```suggestion
				if (TOTAL_ROWS <= PAGE_SIZE) {
```

File: src/main/java/org/isf/medicalinventory/gui/InventoryBrowser.java [hunk 11]
```diff
@@ -274,10 +273,21 @@ private GoodDateChooser getJCalendarTo() {
...
 			jCalendarTo.addDateChangeListener(event -> {
-				LocalDate newDate = event.getNewDate();
-				if (newDate != null) {
-					dateTo = newDate.atTime(LocalTime.MAX);
+				dateTo = jCalendarTo.getDateStartOfDay();
+				if (inventoryList != null) {
+					inventoryList.clear();
 				}
+				InventoryBrowsingModel inventoryModel = new InventoryBrowsingModel();
+				TOTAL_ROWS = inventoryModel.getRowCount();
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/medicalinventory/gui/InventoryBrowser.java:281]: Variables in all caps should only be final static constants and not be changed.

File: src/main/java/org/isf/medicalinventory/gui/InventoryBrowser.java [hunk 12]
```diff
@@ -287,9 +297,18 @@ private GoodDateChooser getJCalendarFrom() {
...
 			jCalendarFrom.addDateChangeListener(event -> {
-				LocalDate newDate = event.getNewDate();
-				if (newDate != null) {
-					dateFrom = newDate.atTime(LocalTime.MAX);				}
+				dateFrom = jCalendarFrom.getDateStartOfDay();
+				InventoryBrowsingModel inventoryModel = new InventoryBrowsingModel();
+				TOTAL_ROWS = inventoryModel.getRowCount();
+				START_INDEX = 0;
+				previous.setEnabled(false);
+				if (TOTAL_ROWS <= PAGE_SIZE){
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/medicalinventory/gui/InventoryBrowser.java:305]:
```suggestion
				if (TOTAL_ROWS <= PAGE_SIZE) {
```

File: bundle/language_en.properties [hunk 13]
```diff
@@ -663,19 +663,39 @@ angal.hospital.thevisitinghourvaluesmustbeintherange0to24.msg
...
 angal.inventory.status.txt                                                                             = Status
 angal.inventory.specificproduct.txt                                                                    = Specific product
 angal.inventoryrow.theorticqty.col                                                                     = Theoritic Qty
+angal.inventory.totalprice                                                                             = Total price
+angal.inventoryrow.unitprice.col                                                                       = Unit price
 angal.inventory.validated                                                                              = Validated
+angal.inventory.validate.btn                                                                           = Validate
+angal.inventory.validate.btn.key                                                                       = V
+angal.inventory.view.btn                                                                               = View
+angal.inventory.view.btn.key                                                                           = E
```
Comment (dbmalkovsky) [inline bundle/language_en.properties:698]: Why not use the common `View` button constants and change the button key for `Validate` instead?

File: src/main/java/org/isf/medicalinventory/gui/InventoryBrowser.java [hunk 14]
```diff
@@ -393,9 +412,27 @@ class InventoryBrowsingModel extends DefaultTableModel {
...
+			String type = InventoryType.principal.toString();
+			try {
+				inventoryList = medicalInventoryManager.getMedicalInventoryByParams(dateFrom, dateTo, state, type);
+			} catch (OHServiceException e) {
+				OHServiceExceptionUtil.showMessages(e);
+			}	
 		}
-
+		
 		public InventoryBrowsingModel(int start_index, int page_size) {
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/medicalinventory/gui/InventoryBrowser.java:425]: The project uses camelCase variables and not those with underscores:  so `startIndex` instead of `start_index`, etc.

File: src/main/java/org/isf/medicalinventory/gui/InventoryBrowser.java [hunk 15]
```diff
@@ -492,4 +551,18 @@ public void InventoryCancelled(AWTEvent e) {
 		}
 		jTableInventory.setModel(new InventoryBrowsingModel());		
 	}
+
+	@Override
+	public void InventoryInserted(AWTEvent e) {
+		if(inventoryList!=null) {
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/medicalinventory/gui/InventoryBrowser.java:557]:
```suggestion
		if (inventoryList! = null) {
```

File: src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java [hunk 16]
```diff
@@ -86,24 +105,62 @@ public class InventoryEdit extends ModalJFrame {
...
 
-	private GoodDateChooser jCalendarTo;
+	public static void removeInventoryListener(InventoryListener listener) {
+		InventoryListeners.remove(InventoryListener.class, listener);
+	}
+
+	private void fireInventoryInserted() {
+		AWTEvent event = new AWTEvent(new Object(), AWTEvent.RESERVED_ID_MAX + 1) {
+
+			/**
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java:131]: Why is this here? It is a useless comment block

File: src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java [hunk 17]
```diff
@@ -615,11 +874,163 @@ private JTextField getCodeTextField() {
...
+			} else {
+				inventoryRowSearchList.add(invRow);
+			}
+		}
+		jTableInventoryRow.updateUI();
+		searchTextField.requestFocus();
+	}
+
+	private Medical chooseMedical(String text) throws OHServiceException {
+		HashMap<String, Medical> medicalMap;
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/medicalinventory/gui/InventoryEdit.java:992]: This should be `Map`; use abstract types for definitions

File: src/main/java/org/isf/medicalinventory/gui/MedicalPicker.java [hunk 18]
```diff
@@ -0,0 +1,327 @@
+package org.isf.medicalinventory.gui;
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/medicalinventory/gui/MedicalPicker.java:1]: Missing license header.

I am stopping making comments; there are just too many things that need changing.

File: src/main/java/org/isf/utils/jobjects/InventoryType.java [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <https://www.gnu.org/licenses/>.
+ */
+package org.isf.utils.jobjects;
+
+
+public enum InventoryType {
+	principal, ward
```
Comment (mwithi) [inline src/main/java/org/isf/utils/jobjects/InventoryType.java:26]: I would adopt `main` and `ward` as principal is mostly used for people and authentication (see JWT doc)

**Review by mwithi — APPROVED** [review]
LGTM. Thanks a lot!
